### PR TITLE
JSON Generation and sped up HTML Report

### PR DESCRIPTION
**Pull request by sshivaji**

### CODE REVIEW

File: src/esbmc/bmc.cpp [hunk 1]
```diff
@@ -102,13 +102,18 @@ void bmct::successful_trace()
...
+  goto_tracet goto_trace;
+
   if (witness_output != "")
   {
-    goto_tracet goto_trace;
     log_progress("Building successful trace");
     /* build_successful_goto_trace(eq, ns, goto_trace); */
     correctness_graphml_goto_trace(options, ns, goto_trace);
   }
+  //  build_goto_trace(eq, smt_conv, goto_trace, is_compact_trace);
```
Comment (rafaelsamenezes) [inline src/esbmc/bmc.cpp:113]: Leftover comment

File: src/esbmc/bmc.cpp [hunk 2]
```diff
@@ -147,6 +152,12 @@ void bmct::error_trace(smt_convt &smt_conv, const symex_target_equationt &eq)
   if (options.get_bool_option("generate-html-report"))
     generate_html_report("1", ns, goto_trace, opt_map);
 
+  if (options.get_bool_option("generate-json-report"))
+    generate_json_report("1", ns, goto_trace, opt_map);
+
+  // if (options.get_bool_option("generate-json-report"))
+  //   generate_json_report("1", ns, goto_trace, opt_map);
```
Comment (rafaelsamenezes) [inline src/esbmc/bmc.cpp:159]: Leftover comment

File: src/esbmc/bmc.cpp [hunk 3]
```diff
@@ -285,6 +296,12 @@ void bmct::report_trace(
   bool term = options.get_bool_option("termination");
   bool show_cex = options.get_bool_option("show-cex");
 
+  // if(options.get_bool_option("generate-json-report") && runtime_solver) {
+  //   goto_tracet goto_trace;
+  //   build_goto_trace(eq, *runtime_solver, goto_trace, true);
+  //   generate_json_report("trace", ns, goto_trace, opt_map);
+  // }
```
Comment (rafaelsamenezes) [inline src/esbmc/bmc.cpp:303]: Leftover comment

File: src/esbmc/bmc.cpp [hunk 4]
```diff
@@ -494,6 +514,12 @@ smt_convt::resultt bmct::run(std::shared_ptr<symex_target_equationt> &eq)
       log_warning("No LTL traces seen, apparently");
   }
 
+  // if(options.get_bool_option("generate-json-report") && eq && runtime_solver) {
+  //   goto_tracet goto_trace;
+  //   build_goto_trace(*eq, *runtime_solver, goto_trace, true);
+  //   generate_json_report("final", ns, goto_trace, opt_map);
+  // }
+
```
Comment (rafaelsamenezes) [inline src/esbmc/bmc.cpp:522]: Leftover comment

File: src/esbmc/bmc.cpp [hunk 5]
```diff
@@ -645,6 +671,9 @@ smt_convt::resultt bmct::run_thread(std::shared_ptr<symex_target_equationt> &eq)
       time2string(symex_stop - symex_start),
       eq->SSA_steps.size());
 
+    // Add this here
+    // Add guards to prevent segfault
```
Comment (rafaelsamenezes) [inline src/esbmc/bmc.cpp:675]: What is this about?

File: src/goto-symex/goto_trace.cpp [hunk 6]
```diff
@@ -360,92 +374,141 @@ void correctness_graphml_goto_trace(
...
+}
+
 void show_goto_trace(
   std::ostream &out,
   const namespacet &ns,
   const goto_tracet &goto_trace)
 {
-  unsigned prev_step_nr = 0;
-  bool first_step = true;
+  try {
```
Comment (rafaelsamenezes) [inline src/goto-symex/goto_trace.cpp:417]: Shouldn't we just propagate the error instead?

File: src/goto-symex/goto_trace.cpp [hunk 7]
```diff
@@ -360,92 +374,141 @@ void correctness_graphml_goto_trace(
...
-            {
-              out << "  " << it.function.as_string();
-              if (it.src->pc->location.is_not_nil())
-                out << " at " << it.src->pc->location << std::endl;
-              else
-                out << std::endl;
+    bool found_violation = false;
+    // json test_data;
+    // test_data["steps"] = json::array();
+    // test_data["status"] = "unknown";
```
Comment (rafaelsamenezes) [inline src/goto-symex/goto_trace.cpp:426]: Leftover comment

File: src/goto-symex/goto_trace.cpp [hunk 8]
```diff
@@ -360,92 +374,141 @@ void correctness_graphml_goto_trace(
...
-    default:
-      assert(false);
-    }
+    // test_data["status"] = found_violation ? "violation" : "success";
+
+    // // Write JSON output
+    // std::ofstream json_out("tests.json");
+    // if (json_out.is_open()) {
+    //   json_out << std::setw(2) << test_data << std::endl;
+    // }
```
Comment (rafaelsamenezes) [inline src/goto-symex/goto_trace.cpp:505]: Leftover comment

File: src/goto-symex/goto_trace.cpp [hunk 9]
```diff
@@ -360,92 +374,141 @@ void correctness_graphml_goto_trace(
...
+    
+  } catch (const std::exception& e) {
+    out << "Error processing trace: " << e.what() << "\n";
+  } catch (...) {
+    out << "Unknown error occurred while processing trace\n";
   }
 }
+
+// End of file
```
Comment (rafaelsamenezes) [inline src/goto-symex/goto_trace.cpp:514]: Leftover comment

File: src/goto-symex/goto_trace.cpp [hunk 10]
```diff
@@ -360,92 +374,141 @@ void correctness_graphml_goto_trace(
...
+      }
+    }
+  }
+}
+
 void show_goto_trace(
   std::ostream &out,
   const namespacet &ns,
   const goto_tracet &goto_trace)
 {
```
Comment (rafaelsamenezes) [inline src/goto-symex/goto_trace.cpp:416]: Can't we just create a new method or add a new parameter for the json output?

File: src/goto-symex/goto_trace.h [hunk 11]
```diff
@@ -93,8 +115,12 @@ class goto_tracet
 public:
   typedef std::list<goto_trace_stept> stepst;
   typedef std::map<std::string, std::string, std::less<std::string>> mid;
-  stepst steps;
+  // stepst steps;
```
Comment (rafaelsamenezes) [inline src/goto-symex/goto_trace.h:118]: Why?

File: src/goto-symex/html.cpp [hunk 12]
```diff
@@ -10,14 +10,33 @@
...
 #include <filesystem>
 #include <util/cmdline.h>
+#include <nlohmann/json.hpp>
+#include <regex>
+#include <array>
+#include <string>
+#include <set>
+#include <vector>
+#include <numeric>
+#include <iostream>
```
Comment (rafaelsamenezes) [inline src/goto-symex/html.cpp:20]: Avoid the use of iostream

File: src/goto-symex/html.cpp [hunk 13]
```diff
@@ -10,14 +10,33 @@
...
+
+using json = nlohmann::json;
+
+static json tests_json;
+static bool first_write = true;
+static std::map<std::string, std::vector<std::string>> source_files;
```
Comment (rafaelsamenezes) [inline src/goto-symex/html.cpp:26]: Avoid global (and static) variables. If you really need it, use a static inline inside a class

File: src/goto-symex/html.cpp [hunk 14]
```diff
@@ -10,14 +10,33 @@
...
+
+std::vector<std::string> include_paths = {
+    "/usr/include",
+    "/usr/local/include"
+    // Add other system include paths
+};
```
Comment (rafaelsamenezes) [inline src/goto-symex/html.cpp:32]: Why this?

File: src/goto-symex/html.cpp [hunk 15]
```diff
@@ -423,17 +459,52 @@ class html_report
   }
 };
 
+// Static member definitions
+std::unordered_map<std::string, std::vector<std::string>> html_report::file_cache;
+std::vector<std::regex> html_report::keyword_patterns;
+bool html_report::patterns_initialized = false;
```
Comment (rafaelsamenezes) [inline src/goto-symex/html.cpp:465]: Static inline at the class

File: src/goto-symex/html.cpp [hunk 16]
```diff
@@ -423,17 +459,52 @@ class html_report
...
+
+const std::vector<std::string>& html_report::get_file_lines(const std::string& filename)
+{
+  auto it = file_cache.find(filename);
+  if(it != file_cache.end()) {
+    return it->second;
+  }
+
+  std::vector<std::string> lines;
+  lines.reserve(500);
```
Comment (rafaelsamenezes) [inline src/goto-symex/html.cpp:487]: Maybe this should be a proper constant elsewhere

File: src/goto-symex/html.cpp [hunk 17]
```diff
@@ -700,3 +736,307 @@ std::string html_report::code_steps::to_html(size_t last) const
     msg,
     next_step_str);
 }
+
+std::string resolve_header_path(const std::string& header, const std::string& base_path) {
```
Comment (rafaelsamenezes) [inline src/goto-symex/html.cpp:740]: I don't think that this module should be the one to sort this out. It should be instances of languaget

File: src/goto-symex/html.cpp [hunk 18]
```diff
@@ -700,3 +736,307 @@ std::string html_report::code_steps::to_html(size_t last) const
...
+
+                    // Add step info
+                    json step_data;
+                    step_data["file"] = file; 
+                    step_data["line"] = line_str;
+                    step_data["function"] = function;
+                    step_data["step_number"] = step_count++;
+
+                    // Add right before the if-else chain for type checking
+                    std::cout << "Step info for " << function << ":" << std::endl;
```
Comment (rafaelsamenezes) [inline src/goto-symex/html.cpp:872]: No cout, use the log interfaces.

File: src/goto-symex/html.cpp [hunk 19]
```diff
@@ -700,3 +736,307 @@ std::string html_report::code_steps::to_html(size_t last) const
...
+            }
+        }
+    }
+    
+    return headers;
+}
+
+void add_coverage_to_json(
+  const goto_tracet &goto_trace,
+  const namespacet &ns)
```
Comment (rafaelsamenezes) [inline src/goto-symex/html.cpp:803]: Maybe add an output stream as a parameter here?

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]
Thanks for the PR. I have a few remarks:

1. This PR is doing too many things: JSON, caching regexes, caching files, figuring out header paths. It makes the reviewing harder.
2. There are lots of left comments, they should be fixed.
3. Be mindful with global variables.
4. Do not use std::cout directly

File: src/esbmc/bmc.cpp [hunk 20]
```diff
@@ -102,13 +102,18 @@ void bmct::successful_trace()
...
   {
-    goto_tracet goto_trace;
     log_progress("Building successful trace");
     /* build_successful_goto_trace(eq, ns, goto_trace); */
     correctness_graphml_goto_trace(options, ns, goto_trace);
   }
+  //  build_goto_trace(eq, smt_conv, goto_trace, is_compact_trace);
+
+   if (options.get_bool_option("generate-json-report"))
+    generate_json_report("1", ns, goto_trace, opt_map);
```
Comment (ChenfengWei0) [inline src/esbmc/bmc.cpp:116]: May I ask if you could add some documentation or comment here?

File: src/goto-symex/goto_trace.cpp [hunk 21]
```diff
@@ -24,6 +33,11 @@ void goto_trace_stept::dump() const
   log_debug("goto-trace", "{}", oss.str());
 }
 
+void goto_trace_stept::track_coverage(goto_tracet const& trace) const
```
Comment (ChenfengWei0) [inline src/goto-symex/goto_trace.cpp:36]: What does this function do?

**Review by ChenfengWei0 — COMMENTED** [review]
Overall LGTM. However, it broke some previous regression tests. For instance,

```
======================================================================
FAIL: test_github_1490 (__main__.RegressionBase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "D:\a\esbmc\esbmc\regression\testing_tool.py", line 249, in test
    self.fail(error_message_prefix + error_message)
AssertionError: 
TEST: D:/a/esbmc/esbmc/regression/esbmc\github_1490
EXPECTED TO FIND: ['State 4 file .* line 9 column 3 function main thread 0']

PROGRAM OUTPUT
ESBMC version 7.7.0 64-bit x86_64 windows
Target: 64-bit little-endian x86_64-unknown-windows with esbmclibc
Parsing D:/a/esbmc/esbmc/regression/esbmc\github_1490\main.c
D:/a/esbmc/esbmc/regression/esbmc\github_1490\main.c:10:2: warning: implicit declaration of function 'assert' is invalid in C99 [-Wimplicit-function-declaration]
        assert(r < 10);
        ^
Converting
Generating GOTO Program
GOTO program creation time: 0.500s
GOTO program processing time: 0.000s
Starting Bounded Model Checking
Symex completed in: 0.000s (21 assignments)
Slicing time: 0.000s (removed 9 assignments)
Generated 1 VCC(s), 1 remaining after simplification (12 assignments)
No solver specified; defaulting to z3
Encoding remaining VCC(s) using bit-vector/floating-point arithmetic
Encoding to solver time: 0.000s
Solving with solver Z3 v4.13.3
Runtime decision procedure: 0.001s
Building error trace

[Counterexample]


[Counterexample]
Violation at: file D:/a/esbmc/esbmc/regression/esbmc\github_1490\main.c line 10 column 2 function main
Reason: assertion


VERIFICATION FAILED
```

Some information in the counterexample is missing somehow.

Also, could you please add some regression tests for this PR?

**Comment by sshivaji** [comment]
I will split this into a few PRs, one for JSON report, one for html speedup
etc and resubmit

Em ter., 5 de nov. de 2024 04:43, Chenfeng Wei ***@***.***>
escreveu:

**Comment by sshivaji** [comment]
I did not program in C++ for a decade, so it was hard work to even get it to this level. I will split the PR accordingly and resubmit :)

**Pull request closed without merging**